src/ssh/allowed_signer.rs

```diff
@@ -0,0 +1,248 @@
+            }
+        }
+
+        Ok(AllowedSigner{
+            principals,
+            cert_authority,
+            namespaces,
+            valid_after,
+            valid_before,
+            key,
+        })
+    }
+}
+
+impl fmt::Display for AllowedSigner {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut output = String::new();
+
+        output.push_str(&self.principals.join(","));
+        
+        if self.cert_authority {
+            output.push_str(" cert-authority");
+        }
+
+        if let Some(ref namespaces) = self.namespaces {
+            output.push_str(&format!(" namespaces={}", namespaces.join(",")));
+        }
+
+        if let Some(valid_after) = self.valid_after {
+            output.push_str(&format!(" valid-after={}", valid_after));
+        }
+
+        if let Some(valid_before) = self.valid_before {
+            output.push_str(&format!(" valid-before={}", valid_before));
+        }
+
+        output.push_str(&format!(" {}", self.key));
+
+        write!(f, "{}", output)
+    }
+}
+
+impl AllowedSigners {
+    /// Reads AllowedSigners from a given path.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use sshcerts::ssh::AllowedSigners;
+    /// fn example() {
+    ///     let allowed_signers = AllowedSigners::from_path("/path/to/allowed_signers").unwrap();
+    ///     println!("{:?}", allowed_signers);
+    /// }
+    /// ```
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<AllowedSigners> {
+        let mut contents = String::new();
+        File::open(path)?.read_to_string(&mut contents)?;
+
+        AllowedSigners::from_string(&contents)
+    }
+
+    /// Parse a collection of allowed signers from a given string.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// use sshcerts::ssh::AllowedSigners;
+    ///
+    /// let allowed_signers = AllowedSigners::from_string(concat!(
+    ///     "user@domain.tld ecdsa-sha2-nistp384 AAAAE2VjZHNhLXNoYTItbmlzdHAzODQAAAAIbmlzdHAzODQAAABhBGJe",
+    ///     "+IDGRhlQdDp+/AIsTXGaVhWaQUbHJwqLDlQIh7V4xatO6E/4Uva+f70WzxgM7xHPGUqafqNAcxxVBP4jkx3HVDRSr7C3",
+    ///     "NVBpr0ZaKXu/hFiCo/4kry4H5MGMEvKATA==\n",
+    ///     "user@domain.tld ecdsa-sha2-nistp384 AAAAE2VjZHNhLXNoYTItbmlzdHAzODQAAAAIbmlzdHAzODQAAABhBGJe",
+    ///     "+IDGRhlQdDp+/AIsTXGaVhWaQUbHJwqLDlQIh7V4xatO6E/4Uva+f70WzxgM7xHPGUqafqNAcxxVBP4jkx3HVDRSr7C3",
+    ///     "NVBpr0ZaKXu/hFiCo/4kry4H5MGMEvKATA==\n"
+    /// )).unwrap();
+    /// println!("{:?}", allowed_signers);
+    /// ```
+    pub fn from_string(s: &str) -> Result<AllowedSigners> {
+        let mut allowed_signers = Vec::new();
+
+        for line in s.split('\n') {
+            let line = line.trim();
+            if line.is_empty() || line.starts_with("#") {
+                continue;
+            }
+            
+            // Remove everything from the first #
+            if let Some(line) = line.split('#').next() {
```

Comment:
This will break for namespaces which contain pound

Reply:
It was intended. But I checked and git's allowed_signers allows # in namespaces. I updated the parsing logic.